auth/src/android/credential_android.cc

```diff
@@ -805,43 +835,158 @@ PhoneAuthProvider::Listener::~Listener() {
-                      j_phone_number, j_time_out, j_milliseconds,
-                      data_->auth_data->app->activity(),
-                      listener->data_->j_listener, j_token);
+  chained_ref =
+      env->CallObjectMethod(builder,
+                            phone_auth_options_builder::GetMethodId(
+                                phone_auth_options_builder::kSetTimeout),
+                            timeout, milliseconds_time_unit);
+  if (util::CheckAndClearJniExceptions(env)) {
+    env->DeleteLocalRef(timeout);
+    env->DeleteLocalRef(milliseconds_time_unit);
+    env->DeleteLocalRef(builder);
+    listener->OnVerificationFailed(
+        "VerifyPhoneNumber: unable set PhoneAuthOptions timeout.");
+    return;
+  }
+  env->DeleteLocalRef(timeout);
+  env->DeleteLocalRef(milliseconds_time_unit);
+  env->DeleteLocalRef(chained_ref);
+
+  // build
+  jobject phone_auth_options =
+      env->CallObjectMethod(builder, phone_auth_options_builder::GetMethodId(
+                                         phone_auth_options_builder::kBuild));
+
+  if (util::CheckAndClearJniExceptions(env)) {
+    env->DeleteLocalRef(builder);
+    listener->OnVerificationFailed(
+        "VerifyPhoneNumber: builder faild to create PhoneAuhtOptions");
+    return;
+  }
+  env->DeleteLocalRef(builder);
+
+  // verify
+  env->CallStaticVoidMethod(
+      phonecred::GetClass(),
+      phonecred::GetMethodId(phonecred::kVerifyPhoneNumber),
+      phone_auth_options);
 
   if (firebase::util::CheckAndClearJniExceptions(env)) {
     // If an error occurred with the call to verifyPhoneNumber, inform the
     // listener that it failed.
-    if (listener) {
-      if (!phone_number || !strlen(phone_number)) {
-        listener->OnVerificationFailed(
-            "Unable to verify with empty phone number");
-      } else {
-        listener->OnVerificationFailed(
-            "Unable to verify the given phone number");
-      }
-    }
+    listener->OnVerificationFailed(
+        "VerifyPhoneNumber: Android to verify the given phone number");
+  }
+
+  env->DeleteLocalRef(phone_auth_options);
+}
+
+void PhoneAuthProvider::VerifyPhoneNumber(
+    const char* phone_number, uint32_t auto_verify_time_out_ms,
+    const ForceResendingToken* force_resending_token, Listener* listener) {
+  FIREBASE_ASSERT_RETURN_VOID(listener != nullptr);
+  JNIEnv* env = GetJniEnv();
+
+  PhoneAuthOptions options;
+  if (force_resending_token != nullptr) {
+    options.force_resending_token =
+        new ForceResendingToken(*force_resending_token);
   }
 
-  env->DeleteLocalRef(j_phone_number);
-  env->DeleteLocalRef(j_milliseconds);
+  if (phone_number != nullptr) {
+    options.phone_number = *phone_number;
+  }
+
+  options.timeout_milliseconds = auto_verify_time_out_ms;
+  VerifyPhoneNumber(options, listener);
+  if (options.force_resending_token != nullptr) {
+    delete options.force_resending_token;
+    options.force_resending_token = nullptr;
+  }
```

Comment:
I'd like to push back on this as the API review has alread been approved. I only have two more days on the team. If this is important than I'd ask you to revise the API proposal and have it re-approved, and then make your suggested coding  change in the feature branch after I've merged this implementation in.

Reply:
Hmm, why don't you just pass the pointer to the option without a new `ForceResendingToken`?

Reply:
Because it's a const pointer and I can't assign it to an options structure which has a non const pointer.

Reply:
I think probably cannot assign a `const ForceResendingToken*` to a field in a struct, not without `const_cast<>`

I would say `const_cast<>` is less of an evil than a new allocation since I think `PhontAuthOptions` is a better in long term.

Reply:
Cleaned it up. PTAL.